torch/_higher_order_ops/cond.py

```diff
@@ -392,28 +392,36 @@ def inner(mode, pred, true_fn, false_fn, operands):
     return trace_cond(mode, cond_op, pred, true_fn, false_fn, operands)
 
 
-@cond_op.py_impl(FakeTensorMode)
-def cond_fake_tensor_mode(mode, pred, true_fn, false_fn, operands):
-    # Ignore here, because if you've gotten here but you're not manually
-    # tracing the inner graphs, that means that you intend to reuse the graph
-    # directly.  Which means the old unbacked symbol bindings are appropriate.
-    # This strategy will not work if unbacked symbols can escape.
-    ignore_fresh_unbacked = contextlib.nullcontext()
-    if mode.shape_env:
-        ignore_fresh_unbacked = mode.shape_env.ignore_fresh_unbacked_symbols()
-
-    with mode, ignore_fresh_unbacked:
-        flat_true_outs, true_out_spec = pytree.tree_flatten(true_fn(*operands))
-        flat_false_outs, false_out_spec = pytree.tree_flatten(false_fn(*operands))
-        if true_out_spec != false_out_spec:
-            raise RuntimeError(
-                "Unmatched output spec from torch.cond branches: "
-                f"true branch tree_spec {true_out_spec} vs false branch tree_spec {false_out_spec}."
-            )
+def _get_fake_mode(inputs: Any) -> FakeTensorMode:
+    fake_mode: Optional[FakeTensorMode] = None
+
+    def _find(t: FakeTensor):
+        nonlocal fake_mode
+        if fake_mode is None:
+            fake_mode = t.fake_mode
+        assert fake_mode is t.fake_mode, (
+            f"Got mixed fake modes {fake_mode}, {t.fake_mode}"
+        )
+
+    pytree.tree_map_only(FakeTensor, _find, inputs)
+    assert fake_mode is not None, "Cannot get fake mode from inputs."
+    return fake_mode
+
+
+@register_fake(cond_op)
```

Comment:
Directly using @cond_op.py_impl(FakeTensorMode) causes us to see DTensor inputs.  register_fake redirects a fake dtensor input to first dispatch to dtensor then fake tensor, in that case fake tensor impl of cond will only see FakeTensor inputs.